File: backend/app/api/api_v1/endpoints/tasks.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
from app.db.database import get_db
from app.services.task import TaskService
from app.schemas.task import (
    CreateTaskRequest, 
    UpdateTaskRequest, 
    TaskResponse, 
    TaskListParams, 
    TaskListResponse,
    TaskStats,
    TaskAction,
    BulkTaskAction,
    BulkTaskResponse,
    CreateTaskTemplateRequest,
    TaskTemplate,
    ScheduledTaskRequest,
    ScheduledTaskResponse
)
from app.core.deps import get_current_active_user as get_current_user
from app.models.user import User

router = APIRouter()


@router.get("/", response_model=TaskListResponse)
async def get_tasks(
    params: TaskListParams = Depends(),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务列表"""
    return await TaskService.get_tasks(db, current_user.id, params)


@router.post("/", response_model=TaskResponse)
async def create_task(
    task_data: CreateTaskRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """创建新任务"""
    return await TaskService.create_task(db, task_data, current_user.id)


@router.get("/stats", response_model=TaskStats)
async def get_task_stats(
    project_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务统计数据"""
    return await TaskService.get_task_stats(db, current_user.id, project_id)


@router.get("/recent")
async def get_recent_tasks(
    limit: int = Query(10, ge=1, le=50),
    project_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取最近任务"""
    params = TaskListParams(
        limit=limit,
        project_id=project_id,
        sort_by="created_at",
        sort_order="desc"
    )
    result = await TaskService.get_tasks(db, current_user.id, params)
    return result.items


@router.get("/running")
async def get_running_tasks(
    project_id: Optional[int] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取运行中的任务"""
    from app.schemas.task import TaskStatus
    params = TaskListParams(
        status=TaskStatus.running,
        project_id=project_id,
        sort_by="started_at",
        sort_order="desc"
    )
    result = await TaskService.get_tasks(db, current_user.id, params)
    return result.items


@router.get("/templates", response_model=list[TaskTemplate])
async def get_task_templates(
    category: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务模板"""
    # TODO: Implement task templates
    raise HTTPException(status_code=501, detail="Task templates not implemented yet")


@router.post("/templates", response_model=TaskTemplate)
async def create_task_template(
    template_data: CreateTaskTemplateRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """创建任务模板"""
    # TODO: Implement task template creation
    raise HTTPException(status_code=501, detail="Task template creation not implemented yet")


@router.post("/templates/{template_id}/create", response_model=TaskResponse)
async def create_task_from_template(
    template_id: int,
    project_id: int,
    name: str,
    description: Optional[str] = None,
    parameters: dict = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """从模板创建任务"""
    # TODO: Implement task creation from template
    raise HTTPException(status_code=501, detail="Task creation from template not implemented yet")


@router.post("/bulk", response_model=BulkTaskResponse)
async def bulk_task_action(
    action_data: BulkTaskAction,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """批量任务操作"""
    return await TaskService.bulk_task_action(db, current_user.id, action_data)


@router.get("/search")
async def search_tasks(
    q: str = Query(..., description="搜索关键词"),
    project_id: Optional[int] = Query(None),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """搜索任务"""
    return await TaskService.search_tasks(db, current_user.id, q, limit)


@router.post("/scheduled", response_model=ScheduledTaskResponse)
async def create_scheduled_task(
    task_data: ScheduledTaskRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """创建定时任务"""
    # TODO: Implement scheduled task creation
    raise HTTPException(status_code=501, detail="Scheduled task creation not implemented yet")


@router.get("/{task_id}", response_model=TaskResponse)
async def get_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取单个任务详情"""
    task = await TaskService.get_task(db, task_id, current_user.id)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.put("/{task_id}", response_model=TaskResponse)
async def update_task(
    task_id: int,
    task_data: UpdateTaskRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """更新任务"""
    task = await TaskService.update_task(db, task_id, current_user.id, task_data)
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.delete("/{task_id}")
async def delete_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """删除任务"""
    try:
        await TaskService.delete_task(db, task_id, current_user.id)
    except HTTPException:
        raise
    return {"message": "Task deleted successfully"}


@router.post("/{task_id}/execute")
async def execute_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """执行任务（触发运行，交由Runner处理）"""
    from app.schemas.task import TaskAction
    action = TaskAction(action="start")
    result = await TaskService.execute_task_action(db, task_id, current_user.id, action)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found or cannot be started")
    return {"status": result.status.value}


@router.post("/{task_id}/action")
async def task_action(
    task_id: int,
    action_data: TaskAction,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """任务操作（启动/取消/确认/重试）"""
    result = await TaskService.execute_task_action(db, task_id, current_user.id, action_data)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found or action not allowed")
    return result


@router.post("/{task_id}/cancel", response_model=TaskResponse)
async def cancel_task(
    task_id: int,
    reason: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """取消任务"""
    from app.schemas.task import TaskAction
    action = TaskAction(action="cancel", reason=reason)
    result = await TaskService.execute_task_action(db, task_id, current_user.id, action)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found or cannot be cancelled")
    return result


@router.post("/{task_id}/retry")
async def retry_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """重试任务"""
    from app.schemas.task import TaskAction
    action = TaskAction(action="retry")
    result = await TaskService.execute_task_action(db, task_id, current_user.id, action)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found or cannot be retried")
    return {"execution_id": result.get("execution_id"), "status": "restarted"}


@router.post("/{task_id}/confirm", response_model=TaskResponse)
async def confirm_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """确认任务（用于需要确认的任务）"""
    from app.schemas.task import TaskAction
    action = TaskAction(action="confirm")
    result = await TaskService.execute_task_action(db, task_id, current_user.id, action)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found or confirmation not needed")
    return result


@router.get("/{task_id}/output")
async def get_task_output(
    task_id: int,
    follow: bool = Query(False),
    tail: int = Query(100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务输出"""
    # TODO: Implement task output retrieval
    raise HTTPException(status_code=501, detail="Task output not implemented yet")


@router.get("/{task_id}/logs")
async def get_task_logs(
    task_id: int,
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
    level: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务日志"""
    # TODO: Implement task logs retrieval
    raise HTTPException(status_code=501, detail="Task logs not implemented yet")


@router.get("/{task_id}/executions")
async def get_task_executions(
    task_id: int,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取任务执行历史"""
    # TODO: Implement task execution history
    raise HTTPException(status_code=501, detail="Task execution history not implemented yet")

File: frontend/src/api/tasks.ts
/**
 * 任务相关API
 */
import { apiClient } from '@/lib/api'

// 类型定义
export interface Task {
  id: number
  project_id?: number
  name: string
  description?: string
  command: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  priority: 'low' | 'medium' | 'high' | 'urgent'
  created_at: string
  updated_at: string
  started_at?: string
  completed_at?: string
  duration?: number // 执行时间（秒）
  output?: string
  error?: string
  exit_code?: number
  progress?: number // 0-100
  metadata?: Record<string, any>
  created_by: string
}

export interface CreateTaskRequest {
  project_id?: number
  name: string
  description?: string
  command: string
  priority?: Task['priority']
  scheduled_at?: string
  schedule?: string
  metadata?: Record<string, any>
}

export interface UpdateTaskRequest {
  name?: string
  description?: string
  priority?: Task['priority']
  scheduled_at?: string
  metadata?: Record<string, any>
}

export interface TaskListParams {
  page?: number
  limit?: number
  project_id?: number
  status?: Task['status'] | Task['status'][]
  priority?: Task['priority']
  search?: string
  sort_by?: 'created_at' | 'updated_at' | 'started_at' | 'completed_at' | 'name' | 'priority'
  sort_order?: 'asc' | 'desc'
  date_from?: string
  date_to?: string
}

export interface TaskListResponse {
  items: Task[]
  total: number
  page: number
  limit: number
  pages?: number
  has_next?: boolean
  has_prev?: boolean
  summary?: {
    pending: number
    running: number
    completed: number
    failed: number
  }
}

export interface TaskExecution {
  id: string
  task_id: string
  status: Task['status']
  started_at: string
  completed_at?: string
  duration?: number
  output: string
  error?: string
  exit_code?: number
}

export interface TaskTemplate {
  id: string
  name: string
  description?: string
  command: string
  category: string
  parameters: Array<{
    name: string
    type: 'string' | 'number' | 'boolean' | 'select'
    required: boolean
    default_value?: any
    options?: string[]
    description?: string
  }>
  created_at: string
}

// 任务API服务类
export class TaskApiService {
  private static readonly BASE_PATH = '/api/v1/tasks'

  /**
   * 获取任务列表
   */
  static async getTasks(params?: TaskListParams): Promise<TaskListResponse> {
    const queryParams = new URLSearchParams()
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          if (Array.isArray(value)) {
            value.forEach(v => queryParams.append(key, v.toString()))
          } else {
            queryParams.append(key, value.toString())
          }
        }
      })
    }

    const url = queryParams.toString() 
      ? `${this.BASE_PATH}?${queryParams.toString()}`
      : this.BASE_PATH

    return apiClient.get<TaskListResponse>(url)
  }

  /**
   * 获取单个任务详情
   */
  static async getTask(id: number | string): Promise<Task> {
    return apiClient.get<Task>(`${this.BASE_PATH}/${id}`)
  }

  /**
   * 创建新任务
   */
  static async createTask(data: CreateTaskRequest): Promise<Task> {
    return apiClient.post<Task>(this.BASE_PATH, data)
  }

  /**
   * 更新任务
   */
  static async updateTask(id: number | string, data: UpdateTaskRequest): Promise<Task> {
    return apiClient.put<Task>(`${this.BASE_PATH}/${id}`, data)
  }

  /**
   * 删除任务
   */
  static async deleteTask(id: number | string): Promise<void> {
    await apiClient.delete(`${this.BASE_PATH}/${id}`)
  }

  /**
   * 执行任务
   */
  static async executeTask(id: number | string): Promise<{
    execution_id: string
    status: string
  }> {
    return apiClient.post(`${this.BASE_PATH}/${id}/execute`)
  }

  /**
   * 取消任务执行
   */
  static async cancelTask(id: number | string): Promise<Task> {
    return apiClient.post<Task>(`${this.BASE_PATH}/${id}/cancel`)
  }

  /**
   * 重新运行任务
   */
  static async retryTask(id: number | string): Promise<{
    execution_id: string
    status: string
  }> {
    return apiClient.post(`${this.BASE_PATH}/${id}/retry`)
  }

  /**
   * 获取任务输出（实时）
   */
  static async getTaskOutput(id: string, params?: {
    follow?: boolean
    tail?: number
  }): Promise<{
    output: string
    is_complete: boolean
    timestamp: string
  }> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/${id}/output?${queryParams.toString()}`
      : `${this.BASE_PATH}/${id}/output`

    return apiClient.get(url)
  }

  /**
   * 获取任务执行历史
   */
  static async getTaskExecutions(id: string, params?: {
    page?: number
    limit?: number
  }): Promise<{
    items: TaskExecution[]
    total: number
    page: number
    limit: number
  }> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/${id}/executions?${queryParams.toString()}`
      : `${this.BASE_PATH}/${id}/executions`

    return apiClient.get(url)
  }

  /**
   * 批量操作任务
   */
  static async batchOperationTasks(
    taskIds: string[],
    operation: 'execute' | 'cancel' | 'delete'
  ): Promise<{
    success: string[]
    failed: Array<{ id: string; error: string }>
  }> {
    return apiClient.post(`${this.BASE_PATH}/batch`, {
      task_ids: taskIds,
      operation
    })
  }

  /**
   * 获取任务统计
   */
  static async getTaskStats(params?: {
    project_id?: number
    date_from?: string
    date_to?: string
  }): Promise<{
    total: number
    pending: number
    running: number
    completed: number
    failed: number
    cancelled: number
    success_rate: number
    by_status?: Record<Task['status'], number>
    by_priority?: Record<Task['priority'], number>
    average_duration?: number
    total_duration?: number
  }> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/stats?${queryParams.toString()}`
      : `${this.BASE_PATH}/stats`

    return apiClient.get(url)
  }

  /**
   * 获取任务模板列表
   */
  static async getTaskTemplates(params?: {
    category?: string
    search?: string
  }): Promise<TaskTemplate[]> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/templates?${queryParams.toString()}`
      : `${this.BASE_PATH}/templates`

    return apiClient.get<TaskTemplate[]>(url)
  }

  /**
   * 根据模板创建任务
   */
  static async createTaskFromTemplate(
    templateId: string,
    data: {
      project_id: string
      name: string
      description?: string
      parameters: Record<string, any>
    }
  ): Promise<Task> {
    return apiClient.post<Task>(`${this.BASE_PATH}/templates/${templateId}/create`, data)
  }

  /**
   * 搜索任务
   */
  static async searchTasks(query: string, params?: {
    project_id?: string
    limit?: number
    filters?: {
      status?: Task['status']
      priority?: Task['priority']
    }
  }): Promise<{
    items: Task[]
    total: number
    query: string
  }> {
    const searchParams = new URLSearchParams({ q: query })
    
    if (params?.project_id) {
      searchParams.append('project_id', params.project_id)
    }
    
    if (params?.limit) {
      searchParams.append('limit', params.limit.toString())
    }
    
    if (params?.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          searchParams.append(`filter_${key}`, value.toString())
        }
      })
    }

    return apiClient.get(`${this.BASE_PATH}/search?${searchParams.toString()}`)
  }

  /**
   * 获取运行中的任务
   */
  static async getRunningTasks(projectId?: string): Promise<Task[]> {
    const params = new URLSearchParams({ status: 'running' })
    if (projectId) {
      params.append('project_id', projectId)
    }

    const response = await this.getTasks({
      status: 'running',
      project_id: projectId,
      sort_by: 'started_at',
      sort_order: 'desc'
    })

    return response.items
  }

  /**
   * 获取最近的任务
   */
  static async getRecentTasks(limit: number = 10, projectId?: string): Promise<Task[]> {
    const response = await this.getTasks({
      limit,
      project_id: projectId,
      sort_by: 'created_at',
      sort_order: 'desc'
    })

    return response.items
  }

  /**
   * 获取失败的任务
   */
  static async getFailedTasks(projectId?: string): Promise<Task[]> {
    const response = await this.getTasks({
      status: 'failed',
      project_id: projectId,
      sort_by: 'updated_at',
      sort_order: 'desc'
    })

    return response.items
  }

  /**
   * 导出任务数据
   */
  static async exportTasks(params?: {
    project_id?: string
    status?: Task['status'][]
    date_from?: string
    date_to?: string
    format: 'csv' | 'json'
  }): Promise<Blob> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          if (Array.isArray(value)) {
            value.forEach(v => queryParams.append(key, v.toString()))
          } else {
            queryParams.append(key, value.toString())
          }
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/export?${queryParams.toString()}`
      : `${this.BASE_PATH}/export`

    const response = await apiClient.getAxiosInstance().get(url, {
      responseType: 'blob'
    })

    return response.data
  }
}

export default TaskApiService